text/0265-event-vision.md

```diff
@@ -0,0 +1,72 @@
+# Event basics
+
+## Motivation
+
+The introduction of Events has been contentious, so we want to write down and agree on a few basics.
+
+### What are events?
+
+Events are a type of log which has a required event name and a specific structure which is implied by that event name.
+
+The event name and its implied structure make events much more suitable for observability compared to traditional logs.
+
+Because of this, the OpenTelemetry project wants to encourage the transition from traditional logs to events.
+
+### OTLP
+
+Since events are a type of log, they share the same OTLP data structure and OTLP pipeline.
+
+### API
+
+OpenTelemetry should have an Event API. This will help to promote the distinction between events and traditional logs,
```

Comment:
> OpenTelemetry should have an Event API

I am unsure. OTel decided to *not* introduce a end-user callable logging API in the past for a reason. Has that reason become invalid now? Or we believe Event API offers something unique that existing logging APIs do not have?

I am more aligned with the wording: "OpenTelemetry may offer an Event API for end users to call", if the existing logging solutions are not suited for generating structured events.

Reply:
Thanks Jack for additional clarification. For the below pseudo-code
```
logger = GetLogger("my-scope");
logger.log(eventname="my-event", severity, body, attributes)
```

Is the idea that, for every event with the name "my-event", the body and attributes will have a fixed schema? And who is enforcing it? In the above, the user can totally produce yet another event with the same name "my-event", but this time with a diff schema for body/attributes. But the same is the case for Event API too:
```
event_logger = GetEventLogger("my-scope");
event_logger.logEvent(eventname="my-event", severity, body, attributes)
```

How does event api enforce that for every "my-event", body/attributes follow the same schema?

Reply:
> Is the idea that, for every event with the name "my-event", the body and attributes will have a fixed schema? And who is enforcing it?

Yes. No enforcement besides the semantic conventions as of now. Its feasible that one day there could be a collector processor which rejects events which don't match the known schemas of an allow list.

> How does event api enforce that for every "my-event", body/attributes follow the same schema?

Folks have speculated that eventually, something like [weaver](https://github.com/open-telemetry/weaver) could be used to generate code which enforces more rigor on the event schema leveraging the languages type system. I.e. weaver could generate an API built on top of the Event API which accepts parameters for the specific fields of the event body.

Reply:
> Is the idea that, for every event with the name "my-event", the body and attributes will have a fixed schema? 

Yes - https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md, specifically
- The event name uniquely identifies event structure / type of the payload (body) and the set of attributes.
- "Unless the same event.name exists on two events, anything in two event bodies is not comparable to each other."

And just because (today) you can "simulate" the sending of an event by including an attribute called `event.name` on a Log Record does not mean (or directly imply) that you should "think" of events as "JUST" a structured Log. An `event` may represent a "thing" that occurred and while it's technically possible to "transport" these via the Logs structure some consumers / applications may want or need to treat these `events` in a different manner to that of a Log record.

Also from the above semantic convention of an event
> Semantically, an Event is a named occurrence at an instant in time. It signals that __"this thing happened at this time"__ and provides additional specifics about the occurrence. Examples of Events might include things like uncaught exceptions, button clicks, user logout, network connection severed, etc.

Reply:
Even if we auto-generate code and instrumentations use them, users still can use the raw API to produce events with same-name-but-diff-schema.

Consider this example where an Event is defined:
```
[LoggerMessage(LogLevel.Critical, "A `{productType}` recall notice was published for `{brandName} {productDescription}` produced by `{companyName}` ({recallReasonDescription}).")]
    public static partial void FoodRecallNotice(
        this ILogger logger,
        string brandName,
        string productDescription,
        string productType,
        string recallReasonDescription,
        string companyName);
```

And it is used as shown below:
```
logger.FoodRecallNotice(
    brandName: "Contoso",
    productDescription: "Salads",
    productType: "Food & Beverages",
    recallReasonDescription: "due to a possible health risk from Listeria monocytogenes",
    companyName: "Contoso Fresh Vegetables, Inc.");
```

This is an example of OTel .NET's existing Log.
1. It has EventName - "FoodRecallNotice"
2. Compiler forces user to pass all the fields, thereby enforcing schema.
3. Still won't user from producing another event of same name with diff. schema!

I believe this is what code-gen tool can potentially generate for each language. For example, Otel Semantic Convention define an Event named `FoodRecallNotice`, it can also generate the helper code to achieve the same. And Instrumentations/End users can chose to use that auto-generated helper method.

Even if the helper is not auto-generated, one can write-it by hand, and ask users to call this to enforce schema.
What would be some of the reasons we ask an end-user to use the Event API via the helper, as opposed to the one I shown above?

Reply:
@cijothomas I was thinking that from .NET perspective, `ILogger` will just provide both - traditional logging and events - it already does, it's just not possible to write structured bodies yet. Same with C++ or  Rust where we could make events first-class citizens and encourage using structured named things.

Reply:
> encourage using structured named things

exactly